bpf/flows.c

```diff
@@ -325,6 +336,66 @@ static inline int flow_monitor(struct __sk_buff *skb, u8 direction) {
     }
     return TC_ACT_OK;
 }
+static inline int export_packet_payload (struct __sk_buff *skb) {
+    void *data_end = (void *)(long)skb->data_end;
+    void *data = (void *)(long)skb->data;
+    payload_meta meta;
+    //packet meta;
+    struct ethhdr *eth  = data;
+    struct iphdr  *ip;
+    struct udphdr *udp_data;
+    
+    
+    
+    if ((void *)eth + sizeof(*eth) > data_end) {
+       return TC_ACT_UNSPEC;	
+    }
+
+    ip = data + sizeof(*eth);
+    if ((void *)ip + sizeof(*ip) > data_end) {
+       return TC_ACT_UNSPEC;	
+    }
+
+    udp_data = (void *)ip + sizeof(*ip);
+    if ((void *)udp_data + sizeof(*udp_data) > data_end) {
+       return TC_ACT_UNSPEC;	
+    }
+
+    if (eth->h_proto != __bpf_htons(ETH_P_IP)) {
+       return TC_ACT_UNSPEC;	
+    }
+
+    //Only analyze UDP packets
+    if (ip->protocol != IPPROTO_UDP ) {
+       return TC_ACT_UNSPEC;	
+    }
+
+    __be16 port = udp_data->dest;
+    //TODO: Update port number/filters to be read from ENV variable
+    __be16 portFromPanoFilter = __bpf_htons(53);
+    bpf_printk("Into export packet payload %d (packet) -- %d\n", port , portFromPanoFilter);
+       
+    if (port == portFromPanoFilter) {
+       meta.if_index = skb->ifindex;
+       meta.pkt_len = data_end - data;
+       meta.payload = data;
+       bpf_perf_event_output(skb, &packet_payloads, ((u64) meta.pkt_len << 32) | BPF_F_CURRENT_CPU, &meta, sizeof(meta));
+    }
+       
+    return TC_ACT_OK;
+
+}
+
+SEC("tc_pano_ingress")
+int ingress_pano_parse (struct __sk_buff *skb) {
+    return export_packet_payload(skb);
+}
+
+SEC("tc_pano_egress")
+int egress_pano_parse (struct __sk_buff *skb) {
+    return export_packet_payload(skb);
+}
+
```

Comment:
we were discussing plans on how to extend the kernel hook 
so the direction was to create folder / the different hook
so we can have some modularity I kinda of started some of this in #106

Reply:
So, this current PR is essentially part of PANO being built on top of the agent.  This is also the reason to write an entire parallel pipeline for this. We can connect over a call if that helps.

Reply:
I also agree that we should perfrom hook point separation in folders, but for this particular PR since the program sections are defined separately they should not get loaded along with our previous hook points.